packages/hash/frontend/src/blocks/userBlocks.tsx

```diff
@@ -0,0 +1,124 @@
+import { BlockMetadata } from "blockprotocol";
+import produce from "immer";
+import {
+  createContext,
+  Dispatch,
+  SetStateAction,
+  useContext,
+  useEffect,
+  useMemo,
+} from "react";
+import { useLocalstorageState } from "rooks";
+
+export type UserBlock = BlockMetadata & {
+  componentId?: string;
+  packagePath?: string;
+};
+
+export const getComponentId = (meta: UserBlock) => {
+  if (meta.componentId) {
+    return meta.componentId;
+  }
+
+  if (meta.packagePath) {
+    return `https://blockprotocol.org/blocks/${meta.packagePath}`;
+  }
+
+  throw new Error("Added a block without packagePath or componentId");
+};
+
+type UserBlocks = UserBlock[];
+
+interface UserBlocksContextState {
+  value: UserBlocks;
+  setValue: Dispatch<SetStateAction<UserBlocks>>;
+}
+
+/** @private enforces use of custom provider */
+const UserBlocksContext = createContext<UserBlocksContextState | null>(null);
+
+const mergeBlocksData = (
+  oldBlocksData: UserBlocks,
+  newBlocksData: UserBlocks,
+): UserBlocks => {
+  return produce(oldBlocksData, (draftUserBlocks) => {
+    for (const latestUserBlock of newBlocksData) {
+      const matchingUserBlockIndex = draftUserBlocks.findIndex(
+        (userBlock) => userBlock.name === latestUserBlock.name,
+      );
+
+      // Using `any` to fix a `Type instantiation is excessively deep` error.
+      // @todo find a potential fix.
+      if (matchingUserBlockIndex === -1) {
+        draftUserBlocks.push(latestUserBlock as any);
+      }
+
+      // Using `any` to fix a `Type instantiation is excessively deep` error.
+      // @todo find a potential fix.
+      if (
+        draftUserBlocks[matchingUserBlockIndex]?.version !==
+        latestUserBlock.version
+      ) {
+        draftUserBlocks[matchingUserBlockIndex] = latestUserBlock as any;
+      }
```

Comment:
`@ts-expect-error` might be better than `as any` here – means you'll be prompted to remove the comment if you fix this at all 

```suggestion
      // @todo Remove need for @ts-expect-error here
      if (matchingUserBlockIndex === -1) {
        // @ts-expect-error TS warns `Type instantiation is excessively deep` but this isn't a problem here
        draftUserBlocks.push(latestUserBlock);
      }

      if (
        draftUserBlocks[matchingUserBlockIndex]?.version !==
        latestUserBlock.version
      ) {
        // @ts-expect-error TS warns `Type instantiation is excessively deep` but this isn't a problem here
        draftUserBlocks[matchingUserBlockIndex] = latestUserBlock;
      }
```